packages/client/src/objectSet/ObjectSet.test.ts

```diff
@@ -436,6 +437,260 @@ describe("ObjectSet", () => {
     });
   });
 
+  describe("Derived Properties Object Set", () => {
+    it("does not allow aggregate or selectProperty before a link type is selected", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          // @ts-expect-error
+          base.aggregate("employeeId:exactDistinct"),
+      });
+
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) =>
+          // @ts-expect-error
+          base.selectProperty("employeeId"),
+      });
+    });
+
+    it("does not allow selectProperty when a many link was selected at any point", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) => {
+          // @ts-expect-error
+          base.pivotTo("peeps").selectProperty("employeeId");
+
+          // @ts-expect-error
+          base.pivotTo("lead").pivotTo("peeps").selectProperty("employeeId");
+
+          return base.pivotTo("lead").selectProperty("employeeId");
+        },
+      });
+    });
+
+    // Executed code fails since we're providing bad strings to the function
+    it.fails("correctly narrows types of aggregate function", () => {
+      client(Employee).withProperties({
+        "derivedPropertyName": (base) => {
```

Comment:
me being annoying nit: but for completeness, lets add an empty string arg as well and make sure that errors